components/script/dom/htmlimageelement.rs

```diff
@@ -44,6 +47,36 @@ impl HTMLImageElement {
     }
 }
 
+// https://html.spec.whatwg.org/multipage/#image-request
+#[derive(JSTraceable, PartialEq, Copy, Clone, HeapSizeOf)]
+pub enum ImageState {
+    Unavailable,
+    PartiallyAvailable,
+    CompletelyAvailable,
+    Broken,
+}
+
+#[must_root]
+#[derive(JSTraceable, HeapSizeOf)]
+pub struct ImageRequest {
+    state: ImageState,
+    current_url: DOMString,
+    image_data: JS<ImageData>,
```

Comment:
Reusing the DOM type here isn't possible, since it stores a JS object and the image data must be accessible off of the thread that runs JS.